### ISSUE TITLE
AutoloadSourceLocator should know all symbols from a file once a class is autoloaded from that file

### ISSUE THREAD

**Issue by ondrejmirtes** [Roave/BetterReflection#617]
Let's say we have a file that looks like this:

```php
class A
{
}

function b()
{
}

const C = 1;
```

Once the PHP runtime autoloads the class `A`, it also knows about `b()` and `C` which traditionally cannot be autoloaded and wouldn't be discovered otherwise.

AutoloadSourceLocator should parse the file with `A` before creating the Reflection and when asked about `b()` and `C`, it should report them as known.

I understand if you don't want this change - it makes the behaviour non-deterministic - if the user asks first about `b()`, it will not be known. But I needed this to make the static reflection's behaviour as close to runtime reflection as possible.

PoC implementation: https://github.com/phpstan/phpstan-src/commit/25c779159e528dfabde191731e6472a123dfecba#diff-b84fb81e70db3cdb6b4495c80b475a44

**Comment by Ocramius** [comment]
Isn't this already handled by looking at loaded files?

I don't understand what is not working as it currently is: specifically test scenario needed.

The `AutoloadSourceLocator` design only follos the **autoloader** behaviour, so I think this might be a new source locator based on a different principle?

**Comment by ondrejmirtes** [comment]
Who is handling loaded files in your opinion? Do you mean the case when `new \ReflectionFunction($functionName)` succeeds and we can read `getFileName()` from it?

Nope, it's not that case - because the function isn't loaded in runtime.

I think AutoloadSourceLocator is a good place to do it because exactly it follows the autoloading principle, and these functions and constants are discovered only thanks to the autoloading of classes in the same file.

Test scenario is straightforward, you have:

1) The file with the class and other symbols (https://github.com/phpstan/phpstan-src/blob/master/tests/PHPStan/Reflection/BetterReflection/SourceLocator/data/a.php)
2) The test case: https://github.com/phpstan/phpstan-src/blob/master/tests/PHPStan/Reflection/BetterReflection/SourceLocator/AutoloadSourceLocatorTest.php

`doFoo`, `SOME_CONSTANT`, and `ANOTHER_CONSTANT` are correctly found only because I asked about `AFoo` first.

I'm not saying if it's a good idea or not, I just wanted to match exactly what PHP does. When `AFoo` is loaded by the autoloader, other symbols in the file are also known.

**Comment by Ocramius** [comment]
The `AutoloadSourceLocator` (and all other locators) should only report what is "seen" during a lookup to a third party responsible of memoization and lookup heuristic/optimisation.

As mentioned here and elsewhere, there is a missing component that requires some architectural design.

Solving the problem in `AutoloadSourceLocator` still leaves it open for all other source locators too.

**Comment by ondrejmirtes** [comment]
Yeah, centralization would make sense. Locators would be able to push all found symbols into it.

**Comment by Ocramius** [comment]
Correct: something like "traverse and stop when found" with a registry that can be set up independently 👍

**Comment by WinterSilence** [comment]
@ondrejmirtes it's unreal/wrong in case like as: 
file `A.php`:
~~~php
class A {}
const C = 1;
~~~
file `B.php`:
~~~
class B {}
const C = 2;
~~~
If we reflect class `A` and `B`, `C` = 1 or 2? or `C` declared in global namespace and it's wrong code?
